lib/api/CDataFrameBoostedTreeRunner.cc

```diff
@@ -106,28 +109,33 @@ CDataFrameBoostedTreeRunner::~CDataFrameBoostedTreeRunner() {
 }
 
 std::size_t CDataFrameBoostedTreeRunner::numberExtraColumns() const {
-    return m_BoostedTree->numberExtraColumnsForTrain();
+    return m_BoostedTreeFactory->numberExtraColumnsForTrain();
 }
 
 void CDataFrameBoostedTreeRunner::writeOneRow(const TStrVec&,
                                               TRowRef row,
                                               core::CRapidJsonConcurrentLineWriter& writer) const {
-    writer.StartObject();
-    writer.Key(PREDICTION);
-    writer.Double(row[m_BoostedTree->columnHoldingPrediction(row.numberColumns())]);
-    writer.EndObject();
+    if (m_BoostedTree) {
+        writer.StartObject();
+        writer.Key(PREDICTION);
+        writer.Double(row[m_BoostedTree->columnHoldingPrediction(row.numberColumns())]);
+        writer.EndObject();
+    } else {
+        HANDLE_FATAL(<< "Boosted tree object was not completely created. Please report this error.");
+    }
 }
 
 void CDataFrameBoostedTreeRunner::runImpl(core::CDataFrame& frame) {
-    m_BoostedTree->train(frame, this->progressRecorder());
+    m_BoostedTree = m_BoostedTreeFactory->frame(frame);
```

Comment:
Suggestion: if we had a public `TBoostedTreeUPtr CBoostedTreeFactory::buildFor(core::CDataFrame& frame)` which replaced both implicit conversion to pointer and `frame` functions then this ensures at compile time that the tree can only be got when frame is non-null and you could ditch the assertion on is `nullptr` in build. I also wonder whether this makes the calling code a little easier to read, i.e. this line would be `m_BoostedTree = m_BoostedTreeFactory->buildFor(frame);` which makes it clearer what is going on IMO.

Reply:
This is actually a very good suggestions! I implemented it.